### PR TITLE
Integrate Pyserini into DSPy

### PR DESCRIPTION
**Pull request by jasper-xian**
Integrate pyserini to enable users to use their own data for retrieval in dspy. I've added a small notebook to demo how it works.

### CODE REVIEW

File: dsp/modules/pyserini.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            topk.append({
+                'text': text,
+                'long_text': text,
+                'pid': pid,
+                'score': hit.score,
+                'rank': rank,
+            })
+
+        if simplify:
+            return [psg["long_text"] for psg in topk]
```
Comment (okhat) [inline dsp/modules/pyserini.py:71]: what is this supposed to do for Pyserini, I think we can remove the whole `simplify` argument

File: dsp/modules/pyserini.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    def __call__(
+        self, query: str, k: int = 10, simplify: bool = False, threads: int = 16,
+    ) -> Union[list[str], list[dotdict]]:
+        hits = self.searcher.search(query, k=k, threads=threads)
+        
+        topk = []
+        for rank, hit in enumerate(hits, start=1):
+            if self.dataset is not None:
+                # Search through user's dataset for the document (could be slow), maybe room for improvement?
```
Comment (okhat) [inline dsp/modules/pyserini.py:52]: what would make it slow? isn't it just a lookup? or does it go to disk?

File: dsp/modules/pyserini.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class PyseriniRetriever:
+    """Wrapper for retrieval with Pyserini. Supports using either pyserini prebuilt faiss indexes or your own faiss index."""
+
+    def __init__(self, 
+                 query_encoder: str = 'castorini/dkrr-dpr-nq-retriever', 
+                 index: str = 'wikipedia-dpr-dkrr-nq', 
+                 dataset: Dataset = None,
+                 id_field: str = 'id',
```
Comment (okhat) [inline dsp/modules/pyserini.py:17]: should the default be 'id' or '_id'?

**Comment by okhat** [comment]
This is *great*. I left three short comments but otherwise I'm happy to merge this!

File: dsp/modules/pyserini.py (hunk 1 above)
Comment (jasper-xian) [reply dsp/modules/pyserini.py:71]: sure, i left it there for the sake of continuity but it's not useful for pyserini (I think), so we can remove it

File: dsp/modules/pyserini.py (hunk 2 above)
Comment (jasper-xian) [reply dsp/modules/pyserini.py:52]: yeah it's just a linear lookup through the dataset

**Comment by ronakice** [comment]
Thanks a lot for this @jasper-xian and @okhat for going through this! Great stuff :)

File: dsp/modules/pyserini.py (hunk 2 above)
Comment (ronakice) [reply dsp/modules/pyserini.py]: Yeah some room for the future improvements!

**Comment by jasper-xian** [comment]
Changes made! I recall we discussed a step-by-step approach to integrating this (first class then indexing, etc.) and it might be useful for you to run the notebook first as a sanity check, but otherwise happy to merge.

Also, maybe I should move the notebook into the docs folder? Up to you @okhat for repo organization.

File: dsp/modules/pyserini.py (hunk 2 above)
Comment (jasper-xian) [reply dsp/modules/pyserini.py]: for the indexible Dataset, it'll be faster to create a map from docid to index during initialization, so maybe we can implement that? @okhat if you have any other ideas let me know as well

**Comment by jasper-xian** [comment]
Also, might be useful to add a blurb about this to the README, perhaps in the Tutorials section?

**Comment by okhat** [comment]
@jasper-xian yes we should add this to the README and we should think more about more "native" indexing, but let's make these two new PRs.

**Comment by okhat** [comment]
The first PR we need after this, though, is Colab runnable notebook. I'll merge your notebook as-is, but have you guys ever ran Pyserini indexing and inference inside Colab? I think it will "just work" since ColBERT just works too.

**Comment by jasper-xian** [comment]
@ronakice can probably provide more insight on running pyserini in Colab, but I'd imagine it should work as is

**Comment by okhat** [comment]
See our own dspy colab notebook. Basically you need an if statement at the top: if this is colab and pyserini is not pip installed, then pip install faiss , pytorch, and pyserini. That’s it really

**Comment by okhat** [comment]
Btw there’s a link pattern to open any github ipynb into colab. See how our url for colab is structured in the readme and just replace that with your own ipynb branch path

File: dsp/modules/pyserini.py (hunk 2 above)
Comment (okhat) [reply dsp/modules/pyserini.py]: It’s a linear lookup? Or constant time lookup?

File: dsp/modules/pyserini.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
+from typing import Union
+import json
+from datasets import Dataset
+
+from dsp.utils import dotdict
+from pyserini.search import FaissSearcher
```
[the remark this answers is not in the document]
Comment (jasper-xian) [reply dsp/modules/pyserini.py]: changes made!

**Comment by jasper-xian** [comment]
I see, I can make the colab changes in a separate PR?

**Comment by jasper-xian** [comment]
Running on Colab right now, getting an error `AttributeError: module 'dspy' has no attribute 'Pyserini'`

I think merging this PR and then doing `git clone` for dspy (same logic as accessing the cache in the intro) should fix this until the next dspy pip release?

**Comment by okhat** [comment]
Merged! Let's get a supplement to this that makes Colab work?

**Comment by jasper-xian** [comment]
yup, on it!

**Comment by jasper-xian** [comment]
Colab PR: #120

**Pull request merged**